### DISCUSSION TITLE
Announcing the tables deprecation on Indexer for Explorer

### DISCUSSION THREAD

**Discussion by telezhnaya** [near/near-indexer-for-explorer#351] [General]
Hey there,
We run [Indexer for Explorer](https://github.com/near/near-indexer-for-explorer) for more than 2 years. All this time we collect the data and serve it to all our users.

We feel responsible for the projects depending on Indexer for Explorer. Each time we've chosen saving the backward compatibility over any breaking improvements.
Unfortunately, we need to admit we can't scale the current Postgres solution infinitely. Due to the size of some tables and spikes in the NEAR blockchain transaction numbers, the Explorer Database starts to lag from the tip of the network, affecting all Explorer users.

As a first compatibility-breaking step for solving this problem, we plan to deprecate 2 tables: `account_changes` and `assets__fungible_token_events`.

## What does table deprecation mean?
Starting from **March 1st, 2023**, we will no longer collect the new data for `account_changes` and `assets__fungible_token_events`. The previous historical data will become unavailable as well.
The source code for these tables will be still available but hidden under the [feature flags](https://github.com/near/near-indexer-for-explorer/pull/349). We do not plan to support these features further.

## What are the workarounds?
Use [Enhanced API](https://console.pagoda.co/apis?tab=enhancedApi#/)! It has much more features and a friendlier API.
Just a part of the highlights: for FTs, we cover legacy tokens such as `aurora` or `wrap.near`. We also show the transaction status together with the transfer info. Finally, we provide you with absolute balances. For native balance, we give more detailed information about any balance changes that may ever occur.

You are still able to run Indexer For Explorer on-premise and continue collecting all the data you want. You need to switch on the corresponding feature flags for that:
```bash
cargo build --release --features "account_changes fungible_token_events"
```

## How to migrate to the new solution?
- For all the standard use cases, [Enhanced API](https://near-mainnet.api.pagoda.co/eapi/v1) works smoothly. You don’t need to validate the input, know the data structures we have under the hood, join tables, etc. Just use the REST API we provided for you. See the [Enhanced API docs](https://console.pagoda.co/apis?tab=enhancedApi#/) for any details.
- If you are sure your workload is pretty common, but it does not fit into [Enhanced API](https://console.pagoda.co/apis?tab=enhancedApi#/), please [ask the question](https://github.com/near/near-enhanced-api-server/issues), we are ready to assist you.
- If you need to run analytical queries or do something unusual, please reach out to our [Self-Hosting section](https://github.com/near/near-indexer-for-explorer#self-hosting) of Indexer for Explorer. 

We apologize for any inconvenience. Our main goal is to provide the community with accurate data with low latency access.

**Comment by telezhnaya** [comment]
FYI we decided to give the community a little more time to migrate. The estimated deprecation date is after April 1st.

**Reply by telezhnaya** [reply]
We finally got rid of these 2 tables.
If you are reading this, probably your prod is down because of these changes. Tag me @telezhnaya, let's think together about how we can help you to migrate faster.

**Reply by iahdevelop** [reply]
Thank you @telezhnaya for the updates. We migrated to Pagoda. But it is really useless. Fore example when we query the transactions history of an address, It didn't return transactionId or receiptId, oldOnwerAccount, etc... It only give you the IN or OUT amount from a certain balance in certain time. Moreover it's not accurate. for some addresses it didn't list any transactions although they have many while for other addresses it only get old transactions. Recent transactions does not appear in Pagoda API calls nor they appear in Near explorer. We are a small company, and we don't afford a hosted indexer right now. We are thinking of dropping the near chain from our App, just because it's so expensive to connect with, compared to other EVM chains for example.  Other EVM Explorers offer a real time reliable data with affordable prices for small projects like us. 
I think for Near to get more adoption, it should offer such a services. 

Thanks.

**Reply by telezhnaya** [reply]
Hey @iahdevelop,
Please provide examples for the points I quoted below.

> Moreover it's not accurate. for some addresses it didn't list any transactions although they have many while for other addresses it only get old transactions.

> Recent transactions does not appear in Pagoda API calls nor they appear in Near explorer.

Please also tell us more about the use case you are interested in. 
I think we can suggest you a solution, but we need more details.
Putting some of my colleagues to cc @pkudinov @khorolets @eduohe @gabehamilton @morgsmccauley hoping we can brainstorm the best solution together

**Reply by khorolets** [reply]
I am doubling this
> Please also tell us more about the use case you are interested in.

We'd like to find a way to help you with your needs, we need to know the needs though.

Also, I am curious about this:
> We are a small company, and we don't afford a hosted indexer right now.

By "hosted indexer" do you mean this project (Indexer for Explorer)? Based on your needs though, usually people don't need the entire Indexer for Explorer (which really quite expensive to run as hosted), and creating and running a custom indexer that covers only your needs would be affordable in cost. 

Looking forward to knowing your needs to be able to suggest you solutions.

**Comment by iahdevelop** [comment]
Thank @telezhnaya and @khorolets  for the reply, Here is an example from Pagoda, 
[2ffa05f73b0dbe8a49e2eb71c2512595abf27eec908b90d47fe2a7f1151efc1c](https://near-mainnet.api.pagoda.co/eapi/v1/accounts/2ffa05f73b0dbe8a49e2eb71c2512595abf27eec908b90d47fe2a7f1151efc1c/balances/FT/token.sweat/history?limit=30)

We made many transactions to this address yesterday, and till now Pagoda didn't reflect any of them. The result looks like this:
`
{
  "history": [],
  "block_timestamp_nanos": "1683185560592087299",
  "block_height": "91062905"
}
`
The same thing is going with the Explorer for example,
[2ffa05f73b0dbe8a49e2eb71c2512595abf27eec908b90d47fe2a7f1151efc1c](https://explorer.mainnet.near.org/accounts/2ffa05f73b0dbe8a49e2eb71c2512595abf27eec908b90d47fe2a7f1151efc1c)

It didn't reflect incoming FT transfers. It only shows IN and OUT Near transfers and only OUT FT transfers. Nonetheless, the transactions took almost 2 hours to appear on the explorer.

Our use case is simple, We have a multichain wallet for micro payments. It's crucial to show the user the transaction history and timely receipts for their transfers on the official explorer as they will use it as proof of payment in their accounting.

Thanks

**Comment by iahdevelop** [comment]
Here is another response from Pagoda [b4322d6ed356b5f2c10db72b7632e17c9d5f7d420a80f06fe751b8dfaa6b38b5](https://near-mainnet.api.pagoda.co/eapi/v1/accounts/b4322d6ed356b5f2c10db72b7632e17c9d5f7d420a80f06fe751b8dfaa6b38b5/balances/FT/token.sweat/history?limit=2)

It lacks transaction_id, receipt_id, new_owner, old_owner, and token_contract_address.

```json5
{
  "history": [
    {
      "event_index": "16750195133876989160000003001000006",
      "involved_account_id": "8ab350cd7fc354c3f1f241fd19d9e1c8b16c8b38ce6ba6abb63300fb67104c3b",
      "delta_balance": "-1020000000000000000",
      "balance": "840000000000000000",
      "cause": "TRANSFER",
      "status": "SUCCESS",
      "metadata": {
        "name": "SWEAT",
        "symbol": "SWEAT",
        "icon": "data:image/svg+xml,%3Csvg viewBox='0 0 100 100' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Crect width='100' height='100' rx='50' fill='%23FF0D75'/%3E%3Cg clip-path='url(%23clip0_283_2788)'%3E%3Cpath d='M39.4653 77.5455L19.0089 40.02L35.5411 22.2805L55.9975 59.806L39.4653 77.5455Z' stroke='white' stroke-width='10'/%3E%3Cpath d='M66.0253 77.8531L45.569 40.3276L62.1012 22.5882L82.5576 60.1136L66.0253 77.8531Z' stroke='white' stroke-width='10'/%3E%3C/g%3E%3Cdefs%3E%3CclipPath id='clip0_283_2788'%3E%3Crect width='100' height='56' fill='white' transform='translate(0 22)'/%3E%3C/clipPath%3E%3C/defs%3E%3C/svg%3E%0A",
        "decimals": 18
      },
      "block_timestamp_nanos": "1675019513387698916",
      "block_height": "84077473"
    },
    {
      "event_index": "16750190831445089700000003001000000",
      "involved_account_id": "8ab350cd7fc354c3f1f241fd19d9e1c8b16c8b38ce6ba6abb63300fb67104c3b",
      "delta_balance": "-1000000000000000000",
      "balance": "1860000000000000000",
      "cause": "TRANSFER",
      "status": "SUCCESS",
      "metadata": {
        "name": "SWEAT",
        "symbol": "SWEAT",
        "icon": "data:image/svg+xml,%3Csvg viewBox='0 0 100 100' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Crect width='100' height='100' rx='50' fill='%23FF0D75'/%3E%3Cg clip-path='url(%23clip0_283_2788)'%3E%3Cpath d='M39.4653 77.5455L19.0089 40.02L35.5411 22.2805L55.9975 59.806L39.4653 77.5455Z' stroke='white' stroke-width='10'/%3E%3Cpath d='M66.0253 77.8531L45.569 40.3276L62.1012 22.5882L82.5576 60.1136L66.0253 77.8531Z' stroke='white' stroke-width='10'/%3E%3C/g%3E%3Cdefs%3E%3CclipPath id='clip0_283_2788'%3E%3Crect width='100' height='56' fill='white' transform='translate(0 22)'/%3E%3C/clipPath%3E%3C/defs%3E%3C/svg%3E%0A",
        "decimals": 18
      },
      "block_timestamp_nanos": "1675019083144508970",
      "block_height": "84077135"
    }
  ],
  "block_timestamp_nanos": "1683186468620591882",
  "block_height": "91063703"
}

**Reply by telezhnaya** [reply]
@iahdevelop thank you for the explanations. There was an incident on our side, now everything is fixed, could you please have a look?

**Reply by telezhnaya** [reply]
> It didn't reflect incoming FT transfers. It only shows IN and OUT Near transfers and only OUT FT transfers. 

It shows both in/out, please check this out
